Zend/zend_operators.c

```diff
@@ -2789,6 +2880,60 @@ ZEND_API zend_string* ZEND_FASTCALL zend_string_tolower_ex(zend_string *str, boo
 }
 /* }}} */
 
+ZEND_API zend_string* ZEND_FASTCALL zend_string_toupper_ex(zend_string *str, bool persistent) /* {{{ */
+{
+	size_t length = ZSTR_LEN(str);
+	unsigned char *p = (unsigned char *) ZSTR_VAL(str);
+	unsigned char *end = p + length;
+
+#ifdef __SSE2__
+	while (p + 16 <= end) {
+		const __m128i _a = _mm_set1_epi8('a' - 1);
+		const __m128i z_ = _mm_set1_epi8('z' + 1);
```

Comment:
These invariant lines should be moved into loop header.

Reply:
This is copied from zend_string_tolower_ex and added in 8ffbd4653e92dbc4d2f142374695a248a94a906c

Maybe this is done in case it's faster for strings of length 0 to 15 not to move the value into a register(?) ahead of time?

Reply:
I was thinking about refactoring the four SSE kernels into a common inline function or macro, so I can make this change to all four as part of that. But note that at -O1 and higher the compiler is already taking the assignment outside the loop.

Reply:
This is done.